crates/task-impls/src/request.rs

```diff
@@ -0,0 +1,261 @@
+    pub private_key: <TYPES::SignatureKey as SignatureKey>::PrivateKey,
+}
+
+/// Alias for a signature
+type Signature<TYPES> =
+    <<TYPES as NodeType>::SignatureKey as SignatureKey>::PureAssembledSignatureType;
+
+impl<TYPES: NodeType, I: NodeImplementation<TYPES>> TaskState for NetworkResponseState<TYPES, I> {
+    type Event = HotShotEvent<TYPES>;
+
+    type Output = HotShotTaskCompleted;
+
+    async fn handle_event(
+        event: Self::Event,
+        task: &mut hotshot_task::task::Task<Self>,
+    ) -> Option<Self::Output> {
+        match event {
+            HotShotEvent::QuorumProposalValidated(proposal) => {
+                let state = task.state();
+                let prop_view = proposal.get_view_number();
+                if prop_view >= state.view {
+                    state.spawn_requests(prop_view, task.clone_sender()).await;
+                }
+                None
+            }
+            HotShotEvent::ViewChange(view) => {
+                if view > task.state().view {
+                    task.state_mut().view = view;
+                }
+                None
+            }
+            HotShotEvent::Shutdown => Some(HotShotTaskCompleted),
+            _ => None,
+        }
+    }
+
+    fn should_shutdown(event: &Self::Event) -> bool {
+        matches!(event, HotShotEvent::Shutdown)
+    }
+    fn filter(&self, event: &Self::Event) -> bool {
+        !matches!(
+            event,
+            HotShotEvent::Shutdown
+                | HotShotEvent::QuorumProposalValidated(_)
+                | HotShotEvent::ViewChange(_)
+        )
+    }
+}
+
+impl<TYPES: NodeType, I: NodeImplementation<TYPES>> NetworkResponseState<TYPES, I> {
+    /// Spawns tasks for a given view to retrieve any data needed.
+    async fn spawn_requests(&self, view: TYPES::Time, sender: Sender<HotShotEvent<TYPES>>) {
+        let requests = self.build_requests(view).await;
+        if requests.is_empty() {
+            return;
+        }
+        requests
+            .into_iter()
+            .for_each(|r| self.run_delay(r, sender.clone(), view));
+    }
+
+    /// Creats the srequest structures for all types that are needed.
+    async fn build_requests(&self, view: TYPES::Time) -> Vec<RequestKind<TYPES>> {
+        let mut reqs = Vec::new();
+        if !self.state.read().await.vid_shares.contains_key(&view) {
+            reqs.push(RequestKind::VID(view, self.public_key.clone()));
+        }
+        // TODO request other things
+        reqs
+    }
+
+    /// run a delayed request task for a request.  The first response
+    /// recieved will be send over `sender`
+    fn run_delay(
+        &self,
+        request: RequestKind<TYPES>,
+        sender: Sender<HotShotEvent<TYPES>>,
+        view: TYPES::Time,
+    ) {
+        let mut recipients: Vec<_> = self
+            .da_membership
+            .get_whole_committee(view)
+            .into_iter()
+            .collect();
+        recipients.shuffle(&mut thread_rng());
```

Comment:
Any change we could get a comment explaining what's going on here? It's not immediately obvious to me why we shuffle the recipients.